docs/architecture/adr-032-typed-events.md

```diff
@@ -0,0 +1,313 @@
+# ADR 032: Typed Events
+
+## Changelog
+
+- 28-Sept-2020: Initial Draft
+
+## Authors
+
+- Anil Kumar (@anilcse)
+- Jack Zampolin (@jackzampolin)
+- Adam Bozanich (@boz) 
+
+## Status
+
+Proposed
+
+## Context
+
+Currently in the SDK, events are defined in the handlers for each message, meaning each module doesn't have a cannonical set of types for each event. Above all else this makes these events difficult to consume as it requires a great deal of raw string matching and parsing. This proposal focuses on updating the events to use **typed events** defined in each module such that emiting and subscribing to events will be much easier. This workflow comes from the experience of the Akash Network team. 
+
+[Our platform](http://github.com/ovrclk/akash) requires a number of programatic on chain interactions both on the provider (datacenter - to bid on new orders and listen for leases created) and user (application developer - to send the app manifest to the provider) side. In addition the Akash team is now maintaining the IBC [`relayer`](https://github.com/ovrclk/relayer), another very event driven process. In working on these core pieces of infrastructure, and integrating lessons learned from Kubernetes developement, our team has developed a standard method for defining and consuming typed events in SDK modules. We have found that it is extremely useful in building this type of event driven application. 
+
+As the SDK gets used more extensively for apps like `peggy`, other peg zones, IBC, DeFi, etc... there will be an exploding demand for event driven applications to support new features desired by users. We propose upstreaming our findings into the SDK to enable all SDK applications to quickly and easily build event driven apps to aid their core application. Wallets, exchanges, explorers, and defi protocols all stand to benefit from this work.
+
+If this proposal is accepted, users will be able to build event driven SDK apps in go by just writing `EventHandler`s for their specific event types and passing them to `EventEmitters` that are defined in the SDK.
```

Comment:
There does seem to be a lot of redundancy I agree, but events are already used pretty widely in the ecosystem and sometimes communicate things not covered in `Msg`s. For instance a cosmwasm contract could have the side effect of creating a gov proposal, even if the `Msg` submitted to the blockchain was just to call a contract.

Reply:
> For instance a cosmwasm contract could have the side effect of creating a gov proposal, even if the Msg submitted to the blockchain was just to call a contract.

If we go with Msg-based OCAPS (i.e. assuming both ADR031 and ADR033 get merged), this contract would need to send a MsgSubmitProposal, which in turn will emit the corresponding Msg Response.

I'd like to see a use-case in this ADR where a message subscription mechanism is not enough.

Reply:
> I'd like to see a use-case in this ADR where a message subscription mechanism is not enough, as I feel ADR031 and ADR033 can get to the same goal in a more elegant way. We could of course still emit events, because of inertia in the ecosystem, but they will be Msg Responses and not typed events.

I don't think it's just inertia. Does the cosmwasm use case I mentioned above not seem like something that isn't handled by `Msg` responses?

Reply:
Possibly with ADR 033 though, we can capture all of the nested `Msg`s that get sent between different modules and maybe it would have the same result...

Reply:
> Possibly with ADR 033 though, we can capture all of the nested Msgs that get sent between different modules and maybe it would have the same result...

Exactly, that's what I tried to convey in the first paragraph of my message.

Reply:
Can someone link to ADR 31 and 33? I don't see open PRs...

Reply:
One thing that isn't covered by `Msg` subscription is the emission of events in `EndBlock`. See the band protocol example: https://blog.cosmos.network/guide-to-building-defi-using-band-protocol-oracle-and-cosmos-ibc-fa5348832f84

Reply:
> If this proposal is accepted, users will be able to build event driven SDK apps

To clarify, these typed events are designed for *external/client-side applications*? I think it is very good to provide the best event support possible for them... working on consuming this since early 2018 and it has been rough getting much working out of it... almost never well tested or structured. I am very happy for such improvements.

If you are saying, as I believe some are interpreting, of using Events to communicate between modules in the SDK, then I have some hesitation. Not that it is bad, but adds a lot of potentials that need to be considered.

Anyway the use of "SDK apps" can be interpretted two ways. I would prefer "SDK modules" or "SDK clients" (and we have strong support for such clients in JS/TS... not just Go.

Reply:
@jackzampolin , could you add a note that subscription to `Msg` doesn't solve all use cases?

Reply:
@ethanfrey this work would also allow for better TS/JS libs around events as well.